File: binary_gap/opussf/bg.py
#!/usr/bin/env python

def solution( n ):
	zeroCount = 0
	oneSeen = False
	zeroCounts = [0]

	while( n > 0 ):
		if( n % 2 == 1 ):
			zeroCounts.append( zeroCount )
			zeroCount = 0
			oneSeen = True
		elif( oneSeen ):
			zeroCount += 1
		n = n >> 1
	return max( zeroCounts )

if __name__=="__main__":
	import unittest

	class testBinaryGap( unittest.TestCase ):
		def setUp( self ):
			pass
		def test01( self ):
			self.assertEquals( solution( 1041 ), 5 )
		def test02( self ):
			self.assertEquals( solution( 1089 ), 5 )
		def test03( self ):
			self.assertEquals( solution( 0 ), 0 )
		def test04( self ):
			self.assertEquals( solution( 20 ), 1 )
		def test05( self ):
			self.assertEquals( solution( 529 ), 4 )
		def test06( self ):
			self.assertEquals( solution( 9 ), 2 )
		def test07( self ):
			self.assertEquals( solution( 8 ), 0 )
		def test08( self ):
			self.assertEquals( solution( 15 ), 0 )
		def test09( self ):
			self.assertEquals( solution( 2147483647 ), 0 )
	unittest.main()

File: binary_gap/opussf/bg.lua
#!/usr/bin/env lua

function solution( n )
	-- return longest binary gap or 0

	-- test for integer
	if( n ~= math.floor( n ) ) then
		return 0
	end
	zeroCount = 0
	zeroCountMax = 0
	oneSeen = false
	while n > 0 do
		bit = n % 2
		if( bit == 1 ) then
			zeroCountMax = math.max( zeroCountMax, zeroCount )
			zeroCount = 0
			oneSeen = true
		elseif( oneSeen ) then
			zeroCount = zeroCount + 1
		end
		--print( bit, zeroCount, zeroCountMax )

		n = math.floor( n / 2 ) -- integer div
	end
	return zeroCountMax
	--if( zeroCountMax > 0 ) then return zeroCountMax; end
end


assert( solution( 1041 ) == 5 )
assert( solution( 1089 ) == 5 )
assert( solution( 0 ) == 0 )
assert( solution( 20 ) == 1 )
assert( solution( 529 ) == 4 )
assert( solution( 9 ) == 2 )
assert( solution( 8 ) == 0 )
assert( solution( 15 ) == 0 )
assert( solution( 5.6 ) == 0 )
assert( solution( 2147483647 ) == 0 )
